certbot-ci/certbot_integration_tests/utils/dns_server.py

```diff
@@ -0,0 +1,136 @@
+#!/usr/bin/env python
+"""Module to setup an RFC2136-capable DNS server"""
+import os
+import os.path
+from pkg_resources import resource_filename
+import shutil
+import socket
+import subprocess
+import sys
+import tempfile
+import time
+
+
+class DNSServer(object):
+    """
+    DNSServer configures and handles the lifetime of an RFC2136-capable server.
+    DNServer provides access to the dns_xdist parameter, listing the address and ports to
+    use for each pytest node.
+
+    At this time, DNSServer should only be used with a single node, but may be expanded in
+    future to support parallelization.
+    """
+
+    BIND_DOCKER_IMAGE = 'internetsystemsconsortium/bind9:9.16'
+    BIND_BIND_ADDRESS = ('127.0.0.1', 45953)
+
+    # A DNS message which is a query for '. IN A' transaction ID 0xe785. This is used by
+    # _wait_until_ready to check that BIND is responding without depending on dnspython.
+    BIND_TEST_QUERY = bytearray.fromhex('0028e7850120000100000000000100000100010000'
+                                        '29100000000000000c000a00083ad084e525a28702')
+
+
+    def __init__(self, nodes, show_output=False):
+        """
+        Create an DNSServer instance.
+        :param list nodes: list of node names that will be setup by pytest xdist
+        :param bool show_output: if True, print the output of the DNS server
+        """
+
+        self.bind_root = tempfile.mkdtemp()
+
+        # Same port to every node. We are only pretending to support parallelization.
+        self.dns_xdist = {
+            node: DNSServer.BIND_BIND_ADDRESS[1] for node in nodes
+        }
+        self.dns_xdist['address'] = DNSServer.BIND_BIND_ADDRESS[0]
+
+        # Unfortunately the BIND9 image forces everything to stderr with -g and we can't
+        # modify the verbosity.
+        self._output = sys.stderr if show_output else open(os.devnull, 'w')
+
+    def start(self):
+        """Start the DNS server"""
+        try:
+            self._configure_bind()
+            self._start_bind()
+        except:
+            self.stop()
+            raise
+
+    def stop(self):
+        """Stop the DNS server, and clean its resources"""
+        try:
+            self.process.terminate()
+            self.process.wait()
+        except:
+            pass
```

Comment:
For troubleshooting purpose maybe it worths a print of what happened if there is a failure.